File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Features/TasksFeature/ViewModels/TasksListViewModel.swift
import Foundation
import SwiftUI
import Combine
import WFMUI

/// ViewModel for tasks list screen
@MainActor
class TasksListViewModel: ObservableObject {
    @Published private(set) var tasks: [Task] = []
    @Published var isLoading = false
    @Published var filterState: TaskState?
    @Published var errorMessage: String? = nil

    private let tasksService: TasksService
    private let userManager: UserManager
    private let toastManager: ToastManager
    private let analyticsService: AnalyticsService
    private var cancellables = Set<AnyCancellable>()
    private var loadTasksTask: _Concurrency.Task<Bool, Never>?

    init(tasksService: TasksService, userManager: UserManager, toastManager: ToastManager, analyticsService: AnalyticsService) {
        self.tasksService = tasksService
        self.userManager = userManager
        self.toastManager = toastManager
        self.analyticsService = analyticsService
        setupEventSubscription()
    }

    /// Подписка на события задач
    private func setupEventSubscription() {
        TaskEventBroadcaster.shared.events
            .sink { [weak self] event in
                guard let self = self else { return }
                _Concurrency.Task {
                    await self.handleTaskEvent(event)
                }
            }
            .store(in: &cancellables)
    }

    /// Обработка события задачи
    private func handleTaskEvent(_ event: TaskEvent) async {
        switch event {
        case .taskCreated, .taskDeleted:
            // Обновить список задач при создании/удалении
            await loadTasks()

        case .taskUpdated(let updatedTask):
            // Обновить задачу локально без запроса на сервер
            if let index = tasks.firstIndex(where: { $0.id == updatedTask.id }) {
                tasks[index] = updatedTask
            }
        }
    }

    /// Вызывается при тапе на карточку задачи — трекает аналитику перед навигацией
    func onTaskCardTapped(_ task: Task) {
        analyticsService.track(.taskCardTapped(
            taskState: task.safeState.rawValue,
            taskType: task.type?.rawValue ?? "PLANNED"
        ))
    }

    /// Load tasks with optional filter
    /// Использует GET /my для получения задач текущей смены
    func loadTasks() async {
        isLoading = true
        await loadTasksInternal()
        isLoading = false
        analyticsService.track(.tasksListViewed(tasksCount: tasks.count))
    }

    /// Refresh tasks list
    ///
    /// Не переключает isLoading в true — это убивает задачу .refreshable в SwiftUI.
    /// Индикатор обновления показывает сам SwiftUI.
    func refresh() async {
        await loadTasksInternal()
    }

    /// Внутренняя логика загрузки задач (без управления isLoading)
    private func loadTasksInternal() async {
        // Отменяем предыдущую загрузку если она еще активна
        loadTasksTask?.cancel()

        // Шаг 1: Проверяем наличие данных
        if let currentUser = userManager.currentUser,
           let currentShift = userManager.currentShift,
           currentShift.status.isActive {
            // Все данные есть - загружаем задачи
            let assignmentId = currentShift.assignmentId
            let success = await loadTasksWithAssignment(assignmentId)

            if success {
                // Задачи загружены успешно
                return
            }

            // Была ошибка при загрузке задач - переходим к шагу 2 (перезапрос UserManager)
        }

        // Шаг 2: Синхронизация с UserManager
        if userManager.isLoading {
            // UserManager загружается - ждём его завершения
            while userManager.isLoading {
                try? await _Concurrency.Task.sleep(nanoseconds: 100_000_000) // 100ms
            }
        } else {
            // UserManager не загружается - сами перезапрашиваем
            await userManager.loadCurrentRole()
        }

        // Шаг 3: Проверяем данные снова
        guard userManager.currentUser != nil else {
            tasks = []
            return
        }

        guard let currentShift = userManager.currentShift,
              currentShift.status.isActive else {
            // Нет смены или смена не открыта — UI покажет EmptyState
            tasks = []
            return
        }

        // Загружаем задачи с полученными данными
        await loadTasksWithAssignment(currentShift.assignmentId)
    }

    /// Загрузка задач с указанным assignment_id
    /// Возвращает true если загрузка успешна, false если была критичная ошибка (нужен перезапрос UserManager)
    @discardableResult
    private func loadTasksWithAssignment(_ assignmentId: Int) async -> Bool {
        // Загружаем задачи с кэшированием (stale-while-revalidate)
        let stream = tasksService.getMyTasks(assignmentId: assignmentId)

        // Создаем Task для итерации по stream
        loadTasksTask = _Concurrency.Task {
            var hadCriticalError = false

            for await result in stream {
                switch result {
                case .cached(let loadedTasks):
                    // Данные из кэша - показываем сразу
                    tasks = loadedTasks
                    errorMessage = nil
                    isLoading = false

                case .fresh(let loadedTasks):
                    // Свежие данные с сервера - обновляем
                    tasks = loadedTasks
                    errorMessage = nil

                case .error(let error):
                    // Ошибка обновления (кэш уже показан если был)
                    if error is CancellationError {
                        // Игнорируем отмену
                        break
                    }

                    // При ошибке от сервера - критичная ошибка, требуется перезапрос UserManager
                    hadCriticalError = true

                    // Показываем сообщение об ошибке если кэша не было
                    if tasks.isEmpty {
                        errorMessage = error.localizedDescription
                    } else {
                        // Кэш был показан, но свежие данные не загрузились
                        if error.shouldShowToUser {
                            toastManager.show(message: error.localizedDescription, state: .error)
                        }
                    }
                }
            }

            return !hadCriticalError
        }

        // Ждем завершения загрузки и возвращаем результат
        return await loadTasksTask?.value ?? false
    }

    /// Get active task for current user (if any)
    var activeTask: Task? {
        tasks.first { $0.state == .inProgress }
    }

    /// Check if user can start a new task (no active tasks)
    var canStartNewTask: Bool {
        activeTask == nil
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/main/java/com/beyondviolet/wfm/features/tasks/presentation/viewmodels/TasksListViewModel.kt
package com.beyondviolet.wfm.features.tasks.presentation.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.beyondviolet.wfm.core.analytics.AnalyticsEvent
import com.beyondviolet.wfm.core.analytics.AnalyticsService
import com.beyondviolet.wfm.core.managers.UserManager
import com.beyondviolet.wfm.core.models.CreateTaskRequest
import com.beyondviolet.wfm.core.models.Task
import com.beyondviolet.wfm.core.models.TaskState
import com.beyondviolet.wfm.core.models.TaskType
import com.beyondviolet.wfm.core.models.isActive
import com.beyondviolet.wfm.core.network.ApiResponse
import com.beyondviolet.wfm.core.network.TasksService
import com.beyondviolet.wfm.ui.components.ToastManager
import com.beyondviolet.wfm.ui.components.WfmToastState
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * ViewModel для списка задач
 */
class TasksListViewModel(
    private val tasksService: TasksService,
    val userManager: UserManager,
    private val toastManager: ToastManager,
    private val analyticsService: AnalyticsService
) : ViewModel() {

    private val _uiState = MutableStateFlow<TasksListUiState>(TasksListUiState.Loading)
    val uiState: StateFlow<TasksListUiState> = _uiState.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _allTasks = MutableStateFlow<List<Task>>(emptyList())

    private var loadTasksJob: Job? = null

    /**
     * Активная задача пользователя (в состоянии IN_PROGRESS)
     * Согласно бизнес-правилу: только 1 активная задача одновременно
     */
    val activeTask: StateFlow<Task?> = _allTasks.map { tasks ->
        tasks.firstOrNull { it.state == TaskState.IN_PROGRESS }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    /**
     * Проверка наличия открытой смены
     * Используется UI для различения:
     * - Нет открытой смены → EmptyState "Список задач будет доступен после открытия смены"
     * - Смена открыта, но задач нет → EmptyState "У вас нет задач"
     */
    val hasOpenShift: StateFlow<Boolean> = userManager.currentShift.map { shift ->
        shift?.status?.isActive() == true
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)

    /**
     * Загружает задачи с учетом текущего фильтра
     * Использует GET /my для получения задач текущей смены
     */
    fun loadTasks() {
        viewModelScope.launch {
            _uiState.value = TasksListUiState.Loading
            loadTasksInternal()
        }
    }

    fun refresh() {
        viewModelScope.launch {
            try {
                _isRefreshing.value = true
                loadTasksInternal()
            } finally {
                // Небольшая задержка чтобы UI успел обработать состояние refresh
                kotlinx.coroutines.delay(200)
                _isRefreshing.value = false
            }
        }
    }

    /**
     * Внутренний метод загрузки задач (используется в loadTasks и refresh)
     */
    private suspend fun loadTasksInternal() {
        // Отменяем предыдущую загрузку если она еще активна
        loadTasksJob?.cancel()

        // Сначала проверяем - может данные уже есть?
        if (userManager.currentUser.value != null &&
            userManager.currentShift.value != null &&
            userManager.currentShift.value?.status?.isActive() == true) {
            // Все данные есть - сразу загружаем задачи!
            android.util.Log.i("TasksListVM", "✅ All data already available, loading tasks for assignment ${userManager.currentShift.value!!.assignmentId}")
            val assignmentId = userManager.currentShift.value!!.assignmentId
            val success = loadTasksWithAssignment(assignmentId)

            if (success) {
                // Задачи загружены успешно
                return
            }

            // Была ошибка при загрузке задач - переходим к шагу 2 (перезапрос UserManager)
            android.util.Log.w("TasksListVM", "⚠️ Error loading tasks, will reload UserManager...")
        }

        // Данных нет или была ошибка - проверяем состояние UserManager
        android.util.Log.w("TasksListVM", "⚠️ Required data missing or error occurred, checking UserManager state...")

        if (userManager.isLoading.value) {
            // UserManager загружается - ждём его завершения
            android.util.Log.i("TasksListVM", "⏳ UserManager is loading, waiting...")
            while (userManager.isLoading.value) {
                kotlinx.coroutines.delay(100) // 100ms
            }
            android.util.Log.i("TasksListVM", "✅ UserManager finished loading")
        } else {
            // UserManager не загружается - сами перезапрашиваем
            android.util.Log.i("TasksListVM", "🔄 UserManager not loading, reloading...")
            userManager.loadCurrentRole()
        }

        // После ожидания/перезапроса проверяем данные снова
        if (userManager.currentUser.value == null) {
            android.util.Log.e("TasksListVM", "❌ User still not loaded after reload, cannot load tasks")
            _allTasks.value = emptyList()
            _uiState.value = TasksListUiState.Success(emptyList())
            return
        }

        val currentShift = userManager.currentShift.value
        if (currentShift == null) {
            // Нет смены — UI покажет EmptyState "Список задач будет доступен после открытия смены"
            android.util.Log.e("TasksListVM", "❌ No shift found after reload, clearing tasks")
            _allTasks.value = emptyList()
            _uiState.value = TasksListUiState.Success(emptyList())
            return
        }

        if (!currentShift.status.isActive()) {
            // Смена не открыта — UI покажет EmptyState "Список задач будет доступен после открытия смены"
            android.util.Log.e("TasksListVM", "❌ Shift not active, clearing tasks")
            _allTasks.value = emptyList()
            _uiState.value = TasksListUiState.Success(emptyList())
            return
        }

        android.util.Log.i("TasksListVM", "✅ Data now available after reload, loading tasks for assignment ${currentShift.assignmentId}")
        loadTasksWithAssignment(currentShift.assignmentId)
    }

    /**
     * Загрузка задач с указанным assignment_id
     * Возвращает true если загрузка успешна, false если была критичная ошибка
     */
    private suspend fun loadTasksWithAssignment(assignmentId: Int): Boolean {
        var hadCriticalError = false

        // Запускаем загрузку задач с кэшированием (stale-while-revalidate)
        loadTasksJob = viewModelScope.launch {
            tasksService.getMyTasks(assignmentId).collect { response ->
                when (response) {
                    is ApiResponse.Success -> {
                        val allTasks = response.data
                        _allTasks.value = allTasks

                        if (response.isCached) {
                            android.util.Log.i("TasksListVM", "📦 Cached tasks loaded: ${allTasks.size} tasks")
                        } else {
                            android.util.Log.i("TasksListVM", "🔄 Fresh tasks loaded: ${allTasks.size} tasks")
                            analyticsService.track(AnalyticsEvent.TasksListViewed(tasksCount = allTasks.size))
                        }

                        _uiState.value = TasksListUiState.Success(allTasks)
                    }

                    is ApiResponse.Error -> {
                        // При ошибке от сервера - критичная ошибка, требуется перезапрос UserManager
                        android.util.Log.e("TasksListVM", "❌ Error loading tasks: ${response.message}")
                        hadCriticalError = true

                        // Показываем toast только если кэша не было
                        if (_allTasks.value.isEmpty()) {
                            toastManager.show(response.message, state = WfmToastState.ERROR)
                        } else {
                            // Кэш был показан, но свежие данные не загрузились
                            if (response.shouldShowToUser()) {
                                toastManager.show(response.message, state = WfmToastState.ERROR)
                            }
                        }
                    }
                }
            }
        }

        // Ждем завершения загрузки и возвращаем результат
        loadTasksJob?.join()
        return !hadCriticalError
    }

    fun onTaskCardTapped(task: Task) {
        analyticsService.track(
            AnalyticsEvent.TaskCardTapped(
                taskState = task.state?.name ?: "UNKNOWN",
                taskType = task.type?.name ?: "UNKNOWN"
            )
        )
    }

    fun createTask(title: String, description: String, plannedMinutes: Int) {
        viewModelScope.launch {
            try {
                val request = CreateTaskRequest(
                    title = title,
                    description = description,
                    type = TaskType.PLANNED,
                    plannedMinutes = plannedMinutes,
                    assigneeId = null,
                    shiftId = null
                )
                when (val response = tasksService.createTask(request)) {
                    is ApiResponse.Success -> loadTasks()
                    is ApiResponse.Error -> {
                        toastManager.show(response.message, state = WfmToastState.ERROR)
                    }
                }
            } catch (e: Exception) {
                toastManager.show(e.message ?: "Ошибка при создании задачи", state = WfmToastState.ERROR)
            }
        }
    }
}

sealed class TasksListUiState {
    data object Loading : TasksListUiState()
    data class Success(val tasks: List<Task>) : TasksListUiState()
    data class Error(val message: String) : TasksListUiState()
}
